src/resource_clients/request_queue.ts

```diff
@@ -96,6 +97,102 @@ export class RequestQueueClient extends ResourceClient {
         return cast(parseDateFields(pluckData(response.data)));
     }
 
+    protected async _batchAddRequestsWithRetries(
+        requests: Array<Omit<RequestQueueClientRequestSchema, 'id'>>,
+        options: RequestQueueClientBatchAddRequestOptions = {},
+    ): Promise<RequestQueueClientBatchAddRequestsResult> {
+        ow(requests, ow.array.ofType(ow.object.partialShape({
+            id: ow.undefined,
+        })).minLength(1).maxLength(25));
+
+        ow(options, ow.object.partialShape({
+            forefront: ow.optional.boolean,
+        }));
+
+        // Keep track of the requests that remain to be processed (in parameter format)
+        let remainingRequests = requests;
+        // Keep track of the requests that have been processed (in api format)
+        let processedRequests: Array<ProcessedRequest> = [];
+        // The requests we have not been able to process in the last call
+        // ie. those we have not been able to process at all
+        let unprocessedRequests: Array<UnprocessedRequest> = [];
+        for (let i = 0; i < 1 + this.httpClient.maxRetries; i++) {
+            const response = await this.httpClient.call({
```

Comment:
Those two layers of retries still confuse me. So there are retries with exponential backoff from the client, but also retries if all the requests were not added successfully, without exponential backoff, right? Which in turn can trigger more retries with exponential backoff.

I'm not sure this will perform optimally under high load. It could trigger hundreds of retries when the rate limits are hit.

Why would some requests be unprocessed? Because the API is under load and can't save the requests fast enough?

Reply:
The unprocessed requests originate in dynamodb. It happens rarely, but sometimes dynamo is not able to save all requests in the batch. We do retries to dynamo with exponential backoff, but it still might happen that we won't be able to store all the requests. Those are sent back as unprocessed (that's in the design of request queue v2, the other option would be treat the whole batch as failed).

On client, there are indeed two levels of retries:
 - the rate limit retries of the client - those do not care about processed / unprocessed items, and the client takes care of exponential backoff
 - the retries for just the unprocessed items returned from api. I thought it would not be needed to have exponential backoff there, just a small delay if any items remain unprocessed